src/profile/utilities.jl

```diff
@@ -0,0 +1,165 @@
+"""
+    TableColumns
+
+A structure containing the column names for the numeric part of the profile table.
+
+The struct also contains a Dict giving the column ranges for Symbols like `:σ` and `:β`.
+Finally it contains a scratch vector used to accumulate to values in a row of the profile table.
+"""
+struct TableColumns{T<:AbstractFloat}
+    cnames::Tuple
+    positions::Dict{Symbol,UnitRange{Int}}
+    v::Vector{T}
+    corrpos::Vector{NTuple{3,Int}}
+end
+
+"""
+    _generatesyms(tag::Char, len::Integer)
+
+Utility to generate a vector of Symbols of the form :<tag><index> from a tag and a length.
+
+The indices are left-padded with zeros to allow lexicographic sorting.
+"""
+function _generatesyms(tag::Char, len::Integer)
+    return Symbol.(string.(tag, lpad.(Base.OneTo(len), ndigits(len), '0')))
+end
+
+function TableColumns(m::LinearMixedModel{T}) where {T}
+    nmvec = [:ζ]
+    positions = Dict(:ζ => 1:1)
+    lastpos = 1
+    sz = m.feterm.rank
+    append!(nmvec, _generatesyms('β', sz))
+    positions[:β] = (lastpos + 1):(lastpos + sz)
+    lastpos += sz
+    push!(nmvec, :σ)
+    lastpos += 1
+    positions[:σ] = lastpos:lastpos
+    sz = sum(t -> size(t.λ, 1), m.reterms)
+    append!(nmvec, _generatesyms('σ', sz))
+    positions[:σs] = (lastpos + 1):(lastpos + sz)
+    lastpos += sz
+    corrpos = NTuple{3,Int}[]
+    for (i, re) in enumerate(m.reterms)
+        (isa(re.λ, Diagonal) || isa(re, ReMat{T,1})) && continue
+        indm = indmat(re)
+        for j in axes(indm, 1)
+            rowj = view(indm, j, :)
+            for k in (j + 1):size(indm, 1)
+                if !iszero(dot(rowj, view(indm, k, :)))
+                    push!(corrpos, (i, j, k))
+                end
+            end
+        end
+    end
+    sz = length(corrpos)
+    if sz > 0
+        append!(nmvec, _generatesyms('ρ', sz))
+        positions[:ρs] = (lastpos + 1):(lastpos + sz)
+        lastpos += sz
+    end
+    sz = length(m.θ)
+    append!(nmvec, _generatesyms('θ', sz))
+    positions[:θ] = (lastpos + 1):(lastpos + sz)
+    return TableColumns((nmvec...,), positions, zeros(T, length(nmvec)), corrpos)
+end
+
+function mkrow!(tc::TableColumns{T}, m::LinearMixedModel{T}, ζ::T) where {T}
+    (; cnames, positions, v, corrpos) = tc
+    v[1] = ζ
+    MixedModels.fixef!(view(v, positions[:β]), m)
+    v[first(positions[:σ])] = m.σ
+    σvals!(view(v, positions[:σs]), m)
+    MixedModels.getθ!(view(v, positions[:θ]), m)  # must do this first to preserve a copy
+    if length(corrpos) > 0
+        ρvals!(view(v, positions[:ρs]), corrpos, m)
+        setθ!(m, view(v, positions[:θ]))
+    end
+    return NamedTuple{cnames,NTuple{length(cnames),T}}((v...,))
+end
+
+function parsej(sym::Symbol) # return the index from symbol names like :θ1, :θ01, etc.
+    symstr = string(sym)                                     # convert Symbol to a String
+    return parse(Int, SubString(symstr, nextind(symstr, 1))) # drop first Unicode character and parse as Int
```

Comment:
There's also [`graphemes`](https://docs.julialang.org/en/v1/stdlib/Unicode/#Unicode.graphemes), which if I recall correctly, appeared in 1.8. (The `compat` entry seems to be missing.)